zerver/lib/message.py

```diff
@@ -845,19 +855,21 @@ def do_render_markdown(
     realm_alert_words_automaton: Optional[ahocorasick.Automaton] = None,
     mention_data: Optional[MentionData] = None,
     email_gateway: bool = False,
-) -> str:
-    """Return HTML for given Markdown. Markdown may add properties to the
-    message object such as `mentions_user_ids`, `mentions_user_group_ids`, and
-    `mentions_wildcard`.  These are only on this Django object and are not
-    saved in the database.
+) -> Tuple[str, RenderedMessage]:
+    """Return HTML and message with additional properties for given Markdown.
+    Markdown may add properties to the message object such as `mentions_user_ids`
+    , `mentions_user_group_ids`, and `mentions_wildcard`.  These are only on this
+    Django object and are not saved in the database. A `RenderedMessage` object is
+    returned to provide a type-safe way to access these properties.
     """
-
-    message.mentions_wildcard = False
-    message.mentions_user_ids = set()
-    message.mentions_user_group_ids = set()
-    message.alert_words = set()
-    message.links_for_preview = set()
-    message.user_ids_with_alert_words = set()
+    # We define msg to explicitly cast message to RenderedMessage to avoid mypy errors
+    rendered_message: RenderedMessage = message
+    rendered_message.mentions_wildcard = False
+    rendered_message.mentions_user_ids = set()
+    rendered_message.mentions_user_group_ids = set()
+    rendered_message.alert_words = set()
+    rendered_message.links_for_preview = set()
+    rendered_message.user_ids_with_alert_words = set()
```

Comment:
I think you need to pass the `RenderedMessage` object to `markdown_convert`, since otherwise the code within there that fills these properties will still be type errors.